### PR TITLE
fix: don't allow overpaying in redeem/refund/replace

### PR DESCRIPTION
**Pull request by sander2**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -24,4 +24,12 @@ members = [
   "primitives",
 ]
 
-resolver = "2"
\ No newline at end of file
+resolver = "2"
+
+[profile.test]
```
Comment (sander2) [inline Cargo.toml:29]: will undo

File: crates/btc-relay/src/types.rs [hunk 2]
```diff
@@ -45,3 +49,85 @@ impl<AccountId, BlockNumber> RichBlockHeader<AccountId, BlockNumber> {
...
+
+pub struct OpReturnPaymentData<T: frame_system::Config> {
+    pub op_return: H256,
+    // vec of (amount, address)
+    payments: Vec<(i64, BtcAddress)>,
+    _marker: sp_std::marker::PhantomData<T>,
+}
+
+impl<T: crate::Config> OpReturnPaymentData<T> {
+    pub fn try_from_transaction(transaction: Transaction) -> Result<Self, DispatchError> {
```
Comment (gregdhill) [inline crates/btc-relay/src/types.rs:61]: Would be better to implement `TryFrom`

File: crates/btc-relay/src/types.rs [hunk 3]
```diff
@@ -45,3 +49,85 @@ impl<AccountId, BlockNumber> RichBlockHeader<AccountId, BlockNumber> {
...
+        // check the number of outputs - this check is redundant due to the checks below, but
+        // this serves to put an upperbound to the number of iterations
+        ensure!(
+            transaction.outputs.len() <= ACCEPTED_MAX_OPRETURN_TRANSACTION_OUTPUTS as usize,
+            Error::<T>::MalformedTransaction
+        );
+
+        let mut payments = Vec::new();
+        let mut op_returns = Vec::new();
```
Comment (gregdhill) [inline crates/btc-relay/src/types.rs:70]: Could adopt an enum here (i.e. `PaymentOrData`), then we already know the expected size of the vector.

File: crates/btc-relay/src/lib.rs [hunk 4]
```diff
@@ -459,10 +449,10 @@ pub const TARGET_TIMESPAN_DIVISOR: u32 = 4;
 pub const ACCEPTED_MIN_TRANSACTION_OUTPUTS: u32 = 1;
 
 // Accepted minimum number of transaction outputs for op-return validation
-pub const ACCEPTED_MIN_TRANSACTION_OUTPUTS_WITH_OP_RETURN: u32 = 2;
+pub const ACCEPTED_MIN_TRANSACTION_OUTPUTS_WITH_OP_RETURN: usize = 2;
 
 // Accepted maximum number of transaction outputs for validation
-pub const ACCEPTED_MAX_TRANSACTION_OUTPUTS: u32 = 32;
+pub const ACCEPTED_MAX_OPRETURN_TRANSACTION_OUTPUTS: u32 = 3;
```
Comment (gregdhill) [inline crates/btc-relay/src/lib.rs:455]:
```suggestion
pub const ACCEPTED_MAX_TRANSACTION_OUTPUTS: u32 = 3;
```

File: crates/btc-relay/src/lib.rs [hunk 5]
```diff
@@ -663,26 +627,82 @@ impl<T: Config> Pallet<T> {
...
+    ) -> Result<(BtcAddress, i64), DispatchError> {
+        let transaction = Self::parse_transaction(&raw_tx)?;
+
+        // Verify that the transaction is indeed included in the main chain
+        Self::_verify_transaction_inclusion(transaction.tx_id(), raw_merkle_proof, None)?;
+
+        Self::get_issue_payment(transaction, recipient_btc_address)
+    }
+
+    fn get_issue_payment(
```
Comment (gregdhill) [inline crates/btc-relay/src/lib.rs:665]: Would prefer `Option<H256>` for generality over specific issue / redeem / replace functions

File: crates/btc-relay/src/types.rs [hunk 6]
```diff
@@ -45,3 +49,85 @@ impl<AccountId, BlockNumber> RichBlockHeader<AccountId, BlockNumber> {
         })
     }
 }
+
+pub struct OpReturnPaymentData<T: frame_system::Config> {
```
Comment (gregdhill) [inline crates/btc-relay/src/types.rs:53]: Not sure that containing this logic adds much over the previous approach

File: crates/btc-relay/src/lib.rs (hunk 5 above)
Comment (sander2) [reply crates/btc-relay/src/lib.rs:665]: I specifically got away from the generality - imo it didn't make sense to generalize, when the input, the output, and the actual checks are all different between issue and [redeem,replace,refund]:



| 			| issue							| redeem/replace/refund                                                              |
| ------------- |------------| -----|
| input:		| address						| 	address, amount, op_return                                                   |
| returns:	| origin_address, paid amount 	| return-to-self address (theft check)                                                             |
| checks:		| any payment to _address_		| payment of correct amount to _address_, exactly 1 correct op_return, at most 1 return-to-self, no unparsable outputs

File: crates/btc-relay/src/types.rs (hunk 6 above)
Comment (sander2) [reply crates/btc-relay/src/types.rs:53]: The advantage is that the theft check and transaction validation logic is now shared, meaning there are no discrepancies between the two.

File: crates/btc-relay/src/types.rs (hunk 3 above)
Comment (sander2) [reply crates/btc-relay/src/types.rs:70]: Are you worried about performance? We could initialize payments to size (num_outputs -1) and op_returns to 1. Having an enum would complicate the logic below

Comment (gregdhill) [reply crates/btc-relay/src/types.rs:70]: I think an enum would improve both performance and readability

File: crates/btc-relay/src/lib.rs (hunk 5 above)
Comment (gregdhill) [reply crates/btc-relay/src/lib.rs:665]: Okay that makes sense

File: crates/btc-relay/src/types.rs (hunk 3 above)
Comment (sander2) [reply crates/btc-relay/src/types.rs:70]: I actually think it'd probably be worse - after creating `PaymentOrData`, you need to iterate over them again to find where the op_return is. Then, removing it from the vec would be `O(n-i)`, and if you don't remove it, you decrease the readability of the rest of code.

Comment (gregdhill) [reply crates/btc-relay/src/types.rs:70]: We could match on the enum vec, but point taken. Just wondering if there's a more streamlined solution here

**Pull request closed without merging**